indra/newview/llfloateremojipicker.cpp

```diff
@@ -672,226 +905,191 @@ void LLFloaterEmojiPicker::onEmojiMouseUp(LLUICtrl* ctrl)
-}
-
-bool LLFloaterEmojiPicker::moveFocusedIconUp()
-{
     for (S32 i = mFocusedIconRow - 1; i >= 0; --i)
     {
         LLScrollingPanel* panel = mEmojiGrid->getPanelList()[i];
         LLEmojiGridRow* row = dynamic_cast<LLEmojiGridRow*>(panel);
-        if (row && row->mList->getPanelList().size() > mFocusedIconCol)
+        if (row && row->mList->getPanelList().size())
         {
             mEmojiScroll->scrollToShowRect(row->getBoundingRect());
+            mFocusedIconCol = row->mList->getPanelList().size() - 1;
             mFocusedIconRow = i;
             selectFocusedIcon();
             return true;
         }
     }
+
     return false;
 }
 
-bool LLFloaterEmojiPicker::moveFocusedIconDown()
+bool LLFloaterEmojiPicker::moveFocusedIconNext()
 {
+    if (mHoveredIcon)
+        return false;
+
+    LLScrollingPanel* panel = mEmojiGrid->getPanelList()[mFocusedIconRow];
+    LLEmojiGridRow* row = dynamic_cast<LLEmojiGridRow*>(panel);
+    S32 colCount = row ? row->mList->getPanelList().size() : 0;
+    if (mFocusedIconCol < colCount - 1)
+    {
+        mFocusedIconCol++;
+        selectFocusedIcon();
+        return true;
+    }
+
     S32 rowCount = mEmojiGrid->getPanelList().size();
     for (S32 i = mFocusedIconRow + 1; i < rowCount; ++i)
     {
         LLScrollingPanel* panel = mEmojiGrid->getPanelList()[i];
         LLEmojiGridRow* row = dynamic_cast<LLEmojiGridRow*>(panel);
-        if (row && row->mList->getPanelList().size() > mFocusedIconCol)
+        if (row && row->mList->getPanelList().size())
         {
             mEmojiScroll->scrollToShowRect(row->getBoundingRect());
+            mFocusedIconCol = 0;
             mFocusedIconRow = i;
             selectFocusedIcon();
             return true;
         }
     }
+
     return false;
 }
 
-bool LLFloaterEmojiPicker::moveFocusedIconLeft()
+void LLFloaterEmojiPicker::selectGridIcon(LLEmojiGridIcon* icon)
 {
-    if (mFocusedIconCol <= 0)
-        return false;
-
-    mFocusedIconCol--;
-    selectFocusedIcon();
-    return true;
+    icon->setBackgroundVisible(TRUE);
+    mPreview->setIcon(icon);
 }
 
-bool LLFloaterEmojiPicker::moveFocusedIconRight()
+void LLFloaterEmojiPicker::unselectGridIcon(LLEmojiGridIcon* icon)
 {
-    LLEmojiGridRow* row = (LLEmojiGridRow*)mEmojiGrid->getPanelList()[mFocusedIconRow];
-    S32 colCount = row->mList->getPanelList().size();
-    if (mFocusedIconCol >= colCount - 1)
-        return false;
-
-    mFocusedIconCol++;
-    selectFocusedIcon();
-    return true;
+    icon->setBackgroundVisible(FALSE);
+    mPreview->setIcon(nullptr);
 }
 
-void LLFloaterEmojiPicker::selectGridIcon(LLUICtrl* ctrl)
+// virtual
+BOOL LLFloaterEmojiPicker::handleKey(KEY key, MASK mask, BOOL called_from_parent)
 {
-    if (LLEmojiGridIcon* icon = dynamic_cast<LLEmojiGridIcon*>(ctrl))
+    if (mask == MASK_NONE)
     {
-        icon->setBackgroundVisible(TRUE);
-        mPreview->setEmoji(icon->getDescr());
+        switch (key)
+        {
+        case KEY_LEFT:
```

Comment:
Ideally this should also handle: home, end, page up, page down.

Reply:
These keys work by default; they scroll the list of emojis up and down without changing the selected emoji